src/metabase/search/impl.clj

```diff
@@ -518,6 +522,63 @@
                                    []))))]
     (map annotate search-results)))
 
+(defn- add-collection-effective-location
+  "Batch-hydrates :effective_location and :effective_parent on collection search results. Keeps search results in
+  order."
+  [search-results]
+  (let [collections    (filter #(mi/instance-of? :model/Collection %) search-results)
+        hydrated-colls (t2/hydrate collections :effective_parent)
+        idx->coll      (into {} (map (juxt :id identity) hydrated-colls))]
+    (map (fn [search-result]
+           (if (mi/instance-of? :model/Collection search-result)
+             (idx->coll (:id search-result))
+             (assoc search-result :effective_location nil)))
+         search-results)))
+
+;;; TODO OMG mix of kebab-case and snake_case here going to make me throw up, we should use all kebab-case in Clojure
+;;; land and then convert the stuff that actually gets sent over the wire in the REST API to snake_case in the API
+;;; endpoint itself, not in the search impl.
+(defn serialize
+  "Massage the raw result from the DB and match data into something more useful for the client"
+  [{:as result :keys [all-scores relevant-scores name display_name collection_id collection_name
+                      collection_authority_level collection_type collection_effective_ancestors effective_parent]}]
+  (let [matching-columns    (into #{} (remove nil? (map :column relevant-scores)))
+        match-context-thunk (first (keep :match-context-thunk relevant-scores))]
+    (-> result
+        (assoc
+         :name           (if (and (contains? matching-columns :display_name) display_name)
+                           display_name
+                           name)
+         :context        (when (and match-context-thunk
+                                    (empty?
+                                     (remove matching-columns search.config/displayed-columns)))
+                           (match-context-thunk))
+         :collection     (merge {:id              collection_id
+                                 :name            collection_name
+                                 :authority_level collection_authority_level
+                                 :type            collection_type}
+                                (when effective_parent
+                                  effective_parent)
```

Comment:
Here's the only real change in this function — override these fields with the fields from `:effective_parent` if it exists (+ dissoc `:effective_parent` below)